src/passes/MemoryCopyFillLowering.cpp

```diff
@@ -0,0 +1,230 @@
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+// Replace memory.copy and memory.fill with a call to a function that
+// implements the same semantics.
+
+namespace wasm {
+struct MemoryCopyFillLowering
+  : public WalkerPass<PostWalker<MemoryCopyFillLowering>> {
+  bool needsMemoryCopy = false;
+  bool needsMemoryFill = false;
+
+  void visitMemoryCopy(MemoryCopy* curr) {
+    assert(curr->destMemory ==
+           curr->sourceMemory); // multi-memory not supported.
+    Builder builder(*getModule());
+    replaceCurrent(builder.makeCall(
+      "__memory_copy", {curr->dest, curr->source, curr->size}, Type::none));
+    needsMemoryCopy = true;
+  }
+
+  void visitMemoryFill(MemoryFill* curr) {
+    Builder builder(*getModule());
+    replaceCurrent(builder.makeCall(
+      "__memory_fill", {curr->dest, curr->value, curr->size}, Type::none));
+    needsMemoryFill = true;
+  }
+
+  void run(Module* module) override {
+    if (!module->features.hasBulkMemory()) {
+      return;
+    }
+    if (module->features.hasAtomics()) {
+      throw "Don't use this pass with atomics"; // alternatively... just don't unset bulkmem feature if atomics? or do nothing?
+    }
+    if (module->features.hasMemory64() || module->features.hasMultiMemory()) {
+      throw "Memory64 and multi-memory not supported"; // TODO: best way to
+                                                       // return an error?
+    }
+    // In order to introduce a call to a function, it must first exist, so
+    // create an empty stub.
+    Builder b(*module);
+    auto memCpyFunc =
+      b.makeFunction("__memory_copy",
+                     {{"dst", Type::i32}, {"src", Type::i32}, {"size", Type::i32}},
+                     Signature({Type::i32, Type::i32, Type::i32}, {Type::none}),
+                     {{"start", Type::i32}, {"end", Type::i32}, {"step", Type::i32}, {"i", Type::i32}});
+    memCpyFunc->body = b.makeBlock();
+    module->addFunction(memCpyFunc.release());
+    auto memFillFunc =
+      b.makeFunction("__memory_fill",
```

Comment:
What happens if a function already exists with this name?  Is there some way to make up a new unique name based on `__memory_fill`?

Also, normally with these passed these string constants get defined up top IIRC.

Reply:
Yeah, making up a new unique name is a good idea. @kripken is there some utility that handles this already?

Reply:
Yes, see e.g.

https://github.com/WebAssembly/binaryen/blob/7a0e738e363d13880ec25018134e178d57c5ba6a/src/tools/fuzzing/fuzzing.cpp#L643